Updated darc documentation (#1951)
dotnet/arcade: Documentation/Darc.md

# Darc 

Darc is a tool for managing and querying the relationships between repositories
in the .NET Core ecosystem. This document describes various scenarios and how to
use darc to achieve them, as well as a general reference guide to darc commands.

## Index
- [Scenarios](#scenarios)
  - [Setting up your darc client](#setting-up-your-darc-client)
  - [Adding dependencies to a repository](#adding-dependencies-to-a-repository)
  - [Updating dependencies in your local repository](#updating-dependencies-in-your-local-repository)
  - [Removing dependencies from a repository](#removing-dependencies-from-a-repository)
  - [Changing a dependency's type](#changing-a-dependencys-type)
  - ['Pinning' dependencies so they do not

>     update](#pinning-dependencies-so-they-do-not-update)

Reviewer: Nit: This can be in the same line

  - [Adding dependency flow](#adding-dependency-flow)
  - [Halting and restarting dependency flow](#halting-and-restarting-dependency-flow)
  - [Viewing the dependency graph](#viewing-the-dependency-graph)
  - [Gathering a build drop](#gathering-a-build-drop)
  
- [Command Reference](#command-reference)
  - [Common Parameters](#common-parameters)
  - [add-channel](#add-channel) - Creates a new channel.
  - [add](#add) - Add a new dependency to Version.Details.xml.
  - [add-default-channel](#add-default-channel) - Add a channel that a build of a branch+repository is automatically applied to.
  - [add-subscription](#add-subscription) - Add a new subscription.
  - [authenticate](#authenticate) - Stores the VSTS and GitHub tokens required for remote operations.
  - [delete-channel](#delete-channel) - Deletes an existing channel.
  - [delete-default-channel](#delete-default-channel) - Remove a default channel association.
  - [delete-subscription](#delete-subscription) - Remove a subscription.

>   - [gather-drop](#gather-drop) - Gather a drop of the outputs for a build

Reviewer: >d [](start = 73, length = 1)

Missing '.'

  - [get-channels](#get-channels) - Get a list of channels.
  - [get-default-channels](#get-default-channels) - Gets a list of repo+branch combinations and their associated default channels for builds.
  - [get-dependencies](#get-dependencies) - Get local dependencies.

>   - [get-dependency-graph](#get-dependency-graph) - Get local dependencies.

Reviewer: >Get local dependencies. [](start = 52, length = 23)

Update (copy/paste from above)

```

After your client is installed, you should be able to launch it by typing 'darc'
on the command line.  If this is the first global tool you've installed, you may
need to first restart your command window.

```
PS C:\enlistments\arcade> darc
Microsoft.DotNet 1.1.0-beta.19081.1+270fa76db13d4c103a6dec2b03f1fd79730ff429
c Microsoft Corporation. All rights reserved.
ERROR(S):
No verb selected.

  add-channel                  Creates a new channel.

  add                          Add a new dependency to Version.Details.xml.

  add-default-channel          Add a channel that a build of a branch+repository is automatically applied to.

  add-subscription             Add a new subscription.

  authenticate                 Stores the VSTS and GitHub tokens required for remote operations.

  delete-channel               Deletes an existing channel.

  delete-default-channel       Remove a default channel association.

...
```

When executing most operations, the client needs to make some remote queries.
These remote queries require authentication in most circumstances. There are 3
PATs that may be used:

> - A GitHub PAT for downloading files off github (e.g. eng/Version.Details.xml or

Reviewer 2: ```suggestion
- A GitHub PAT for downloading files off GitHub (e.g. eng/Version.Details.xml or
```

  arcade script files. 

> - An Azure DevOps pat for downloading files off of Azure DevOps. (e.g.

Reviewer: >pat [](start = 18, length = 3)

pat -> PAT

These tokens can either be passed on the command line using parameters (see [Common parameters](#common-parameters)), or
cached locally on the machine using the [`darc authenticate`](#authenticate) command.

After supplying your secrets, a simple `darc get-channels` operations should succeed.

```
PS C:\enlistments\arcade> darc get-channels
.NET Tools - Latest
.NET Core 3 Dev
.NET Engineering Services - Int
.NET Engineering Services - Prod
.NET Tools - Validation
```

### Adding dependencies to a repository

Adding a new dependency to a repository means adding a dependency that is
tracked by the Maestro++/darc system. This dependency's version can then be automatically updated,
along with its sha and repository uri. It can contribute to the repository
dependency graph, and the build producing it will be included in a drop.

Dependencies are tracked in eng/Version.Details.xml, along with the following
information about each one:
- Name
- Source sha
- Source repository
- Is the dependency version pinned (can be it automatically updated?)
- Dependency type (toolset or product)

The [`darc add`](#add) command adds a new dependency.  It takes a number of

> parameters, though only `--name` and `--type` are initially required.  It is

Reviewer 2: What makes the repo parameter highly recommended?

Consider rewording like "The `--repo` parameter is optional but highly recommended to improve traceability."

Author: Without --repo you can't immediately run --update-dependencies.  I'll rework this

corresponding PropertyGroup entry in eng/Versions.props will be added, with the same
version number. The property name is derived off of the dependency name, with
the .'s and -'s removed and "PackageVersion" added onto the end. *Note: If
eng/Versions.props's existing version property names end with the suffix
'Version', darc will append that instead.*

eng/Versions.props after add command.
```
<MicrosoftNETCoreAppPackageVersion></MicrosoftNETCoreAppPackageVersion>
```

After doing this, you can use the generated `MicrosoftNETCoreAppPackageVersion`
property as inputs to a PackageReference element or wherever else may need a
version number.

After addding, it is recommended that you use darc to fill out the missing
dependency information.  The information can also be filled in after commiting by Maestro++
using dependency flow, but passing your CI without a filled-in version number
may be difficult. See (Updating dependencies in your local repository).

#### Toolset vs. Product Dependencies

There are two types of dependencies: 'Product' and 'Toolset'. Choosing between
them involves answering the question:

> "Does my repository repackage the input dependency's binaries, or information

Reviewer 2: Which answer goes with which option?

    </Dependency>
    <Dependency Name="Microsoft.DotNet.SignTool" Version="1.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.Helix.Sdk" Version="2.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
  </ToolsetDependencies>
</Dependencies>
```

### Adding dependency flow

Dependency flow is the automatic movement of dependency information between
repositories. This is done by Maestro++ keeping track of new builds in each
repository and opening PRs in other repositories when those repositories' input
dependencies in eng/Version.Details.xml match the outputs of the new builds.

These updates are selective and based on 'subscriptions'. A subscription describes an update
operation for a specific repository+branch combination, mapping outputs of a
repository that have beeen applied to a channel (virtual branch) onto matching
inputs of the target repository+branch.

For example, a build of dotnet/corefx might be applied to the ".NET Core 3 Dev"

> channel. dotnet/core-setup maps new outputs of corefx on the ".NET Core 3 Dev"

Reviewer: >corefx [](start = 47, length = 6)

dotnet/corefx

channel onto its master branch.

A subscription has a few parts:
- Mapping of source repo + source channel => target repo target branch
- An update rate (e.g. every day, every build, not at all)
- Whether a subscription is batchable or not. If batchable, all batchable

>   subscriptions targeting the same branch/repo combination will share a PR.

Reviewer: >branch/repo [](start = 35, length = 11)

repo/branch?

>   *Note: Batchable subscriptions are currently unsupported in darc*

Reviewer 2: Describing features that do not yet exist just adds maintenance liability. Better to not mention it at all until after it is released.

Author: JC is working on it right now I think.

Reviewer: Will come after updates which is mostly done :)

---
In reply to: [253236648](https://github.com/dotnet/arcade/pull/1951#discussion_r253236648) [](ancestors = 253236648)

- A set of auto merge policies, if the subscription is not batchable.  If batchable,
  merge policies are set on a repository level rather than a per-subscription
  level, as they end up shared between several subscriptions. *Note: repository
  merge policies are currently unsupported in darc*

For additional information and samples, see [add-subscription](#add-subscription)

> #### When should a subscription be used to update my repository's inputs

Reviewer: >s [](start = 71, length = 1)

Add '?' at the end

Whenever you want regular dependency flow between two repositories. For example,
every day (when there is a new source build) or on every new build.

#### What input channel should be used?

There are generally two channels for day to day use:
- '.NET Core 3 Dev' - Day to day builds of .NET Core 3 repositories are placed on
  this channel.

> - '.NET Tools - Latest' - LKG Arcade builds are placed on this channel.

Reviewer 2: ```suggestion
- '.NET Tools - Latest' - Arcade releases are placed on this channel.
```

For a rough example, let's say that at the end of the shipping cycle core-setup
flows its 'last' build to core-sdk and other repos.  The graph stabilizes.
Then, at the last moment, they make a test only fix and check it in.  It flows
to core-sdk quickly, but gets held up in the other repos that pull
Microsoft.NETCore.App due to them not taking any new PRs.  This certainly counts
as an incoherency, but becuase there is no semantic product difference between
the two versions in the graph there may be no harm in shipping with that
incoherency.

### Gathering a build drop

Gathering a drop of the product or subpiece part of the product is a typical
activity done by developers, release automation, etc.  The darc tool can do this
using the Build Asset Registry information alongside the dependency graph
information.

A build drop is a gathering of all the outputs of a specific build that were
reported to the Build Asset Registry into a local location. Optionally, this
build drop may also include any builds that were inputs to this build, based on
the dependency information in Version.Details.xml. Builds produce various

> outputs that go to various locations. For example, they may produce nuget

Reviewer: >nuget [](start = 68, length = 5)

nuget -> NuGet

> packges, zips, msis, etc. These may be available in the build artifacts, or may

Reviewer: >msis [](start = 15, length = 4)

msis -> MSIs

Reviewer: Both of the above also apply to line 1027

---
In reply to: [253530424](https://github.com/dotnet/arcade/pull/1951#discussion_r253530424) [](ancestors = 253530424)

be located in various storage accounts.  Gather-drop mines the Build Asset
Registry for these locations and downloads them all to the local directory.

The drop is generally divided into two sections: shipping and non-shipping.
Shipping dependencies are those that must be visible to customers.  Non-shipping
dependencies are those that generally serve as inter-repo transport.  Some
examples:
- Packages shipping to nuget.org - Shipping
- Coreclr transport packages - Nonshipping, since they get repackaged by
  core-setup
- Symbols - Shipping, since they go to symbol servers

#### What can I gather a drop for?

> You can gather a either an isolated or full (includes dependencies) drop for any

Reviewer: > a  [](start = 14, length = 3)

I think this 'a' is not needed

build in the repository graph. Since the major components of the .NET Core stack
are rooted at 3 different repositories, gathering a drop of those repositories
will generate a drop for those 3 components and any of their inputs if `--full`
is supplied.
- https://github.com/dotnet/core-sdk - SDK
- https://github.com/aspnet/AspNetCore - ASP.NET Core
- https://github.com/dotnet/core-setup - Runtime.

#### How do I find the root build for a drop?

`gather-drop` must be able to find a root build. This root build today is
identified by either a repo+sha combo or a direct BAR build ID. For most
repositories that build on every commit, providing the repo uri and latest sha
will generate a drop:

```
PS C:\enlistments\core-sdk> git rev-parse HEAD
465a336c7a5ca3af2f6cf5172ddc0ebde620803b

> PS C:\enlistments\core-sdk> darc gather-drop --repo https://github.com/dotnet/core-sdk --commit 465a336c7a5ca3af2f6cf5172ddc0ebde620803b --output-dir C:\scratch\drop

Reviewer: >PS C:\enlistments\core-sdk>  [](start = 0, length = 28)

Split this line in two so the length matches the rest of the doc

...
```

If no build exists at that drop, darc will show an error. In this case, you
might try other recents shas, or use the BAR swagger API
(https://maestro-prod.westus2.cloudapp.azure.com/swagger) to look up a build
id. Remember to authenticate using a token from
https://maestro-prod.westus2.cloudapp.azure.com/.  Better methods of obtaining
the root build are coming soon.

The root build can then be provided using --id

```
PS C:\enlistments\core-sdk> darc gather-drop --id 1234 --output-dir C:\scratch\drop
```

## Command Reference

### **`Common parameters`**

There are a few common parameters available on every command:

- `-p, --password` - Build Asset Registry password.  You can obtain this
  password by going to https://maestro-prod.westus2.cloudapp.azure.com/, logging
  in using the link in the top right, then generating a token using the menu in
  the top right.  This setting overrides whatever BAR password was provided when
  doing `darc authenticate`.

> - `--github-pat` - Person access token used to authenticate GitHub. This is a GitHub PAT used

Reviewer: >Person [](start = 19, length = 6)

Person -> Personal

  to avoid rate limiting when accessing github to download arcade script files
  or version files. You only need a GitHub PAT with **no** authorization scopes

>   checked. This setting overrides whatever BAR password was provided when

Reviewer 2: ```suggestion
  checked. This setting overrides whatever GitHub PAT was provided when
```

  doing `darc authenticate`.
- `--azdev-pat` - Personal access token used to authenticate to Azure DevOps.

>   This token should have Code Read permissions. This setting overrides whatever BAR password was provided when

Reviewer 2: ```suggestion
  This token should have Code Read permissions. This setting overrides whatever Azure DevOps PAT was provided when
```

  doing `darc authenticate`.
- `--bar-uri` - URI of the build asset registry service to use.  Typically left

>   as its default (https://maestro-prod.westus2.cloudapp.azure.com) This setting overrides whatever BAR password was provided when

Reviewer 2: Text on "BAR password" seems like a copy-paste error?

  doing `darc authenticate`.
- `--verbose` - Turn on additional output.
- `--debug` - Turn on debug output
- `--help` - Display help
- `--version` - Display version of darc.

Individual darc commands are described below.

### **`add-channel`**

Add a new channel. This creates a new tag that builds can be applied to.

*This is not a typical operation and you should consult with the (`@dnceng`)
engineering team before doing so.*

**Sample**:
```
PS D:\enlistments\arcade> darc add-channel --name "Foo"

Successfully created new channel with name 'Foo'.
```

**Parameters**

- `-n, --name` -  **(Required)**. Name of channel to create.
- `-c, --classification` - Classification of channel. Defaults to 'dev'.  Today,
  this classification does not affect any functionality
- `-i, --internal` - Channel is internal only. This option is currently
  non-functional

**See also**:
- [delete-channel](#delete-channel)
- [get-channels](#get-channels)

### **`add`**

> Add a new tracked dependency to the Version.Detail.xml file in your local repo.

Reviewer: >Add a new tracked dependency to the Version.Detail.xml [](start = 0, length = 54)

'add' also updates Versions.props and global.json depending on the added dependency. Might be worth mentioning that

<Dependencies>
  <ProductDependencies></ProductDependencies>
  <ToolsetDependencies>
    <Dependency Name="Microsoft.DotNet.Arcade.Sdk" Version="1.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.Build.Tasks.Feed" Version="2.2.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.Maestro.Tasks" Version="1.0.0-beta.19060.8">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>67384d20d310611afc1c2b4dd3b953fda182def4</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.SignTool" Version="1.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.Helix.Sdk" Version="2.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
  </ToolsetDependencies>
</Dependencies>
```

Running add

```

> PS D:\enlistments\arcade> darc add --name "Microsoft.NETCore.App" --type "product" --version 1 --commit 2 --repo https://github.com/dotnet/core-setup

Reviewer: >PS D:\enlistmen [](start = 0, length = 15)

Split in two

    </Dependency>
    <Dependency Name="Microsoft.DotNet.Maestro.Tasks" Version="1.0.0-beta.19060.8">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>67384d20d310611afc1c2b4dd3b953fda182def4</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.SignTool" Version="1.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
    <Dependency Name="Microsoft.DotNet.Helix.Sdk" Version="2.0.0-beta.19080.6">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>14d1133b6074b463784a7adbbf385df0462f4010</Sha>
    </Dependency>
  </ToolsetDependencies>
</Dependencies>
```

**See also**:
- [update-dependencies](#update-dependencies)
- [Adding dependencies to a repository](#adding-dependencies-to-a-repository)

### **`add-default-channel`**

Adds a new default channel mapping.  A default channel maps each new build of a specific
branch of a repository onto a specific channel. While builds can be selectively
and manually applied to channels, this is generally inconvenient for day to day development
in most cases.  In general, until release shutdown, each build of a branch

> should always be applied to its "normal" channel.,

Reviewer: >, [](start = 49, length = 1)

Remove ','

- `--branch` - **(Required)** Build of 'repo' on this branch will be
  automatically applied to 'channel'.  Should generally be "refs/heads/branchName"
- `--repo` - **(Required)** Build of this repo repo on 'branch' will be automatically applied to 'channel'

**Sample**
```
PS D:\enlistments\arcade> darc add-default-channel --channel ".Net Core 3 Dev" --branch refs/heads/master --repo https://github.com/dotnet/arcade
```

**See also**:
- [get-channels](#get-channels)
- [get-default-channels](#get-default-channels)
- [delete-default-channel](#delete-default-channel)

### **`add-subscription`**

Adds a new subscription to Maestro++.

A subscription describes an update
operation for a specific repository+branch combination, mapping outputs of a
repository that have beeen applied to a channel (virtual branch) onto matching
inputs of the target repository+branch.

For example, a build of dotnet/corefx might be applied to the ".NET Core 3 Dev"
channel. dotnet/core-setup maps new outputs of corefx on the ".NET Core 3 Dev"
channel onto its master branch.

A subscription has a few parts:

> - Mapping of source repo + source channel => target repo target branch

Reviewer: >target repo target branch [](start = 45, length = 25)

target repo + target branch

- An update rate (e.g. every day, every build, not at all)
- Whether a subscription is batchable or not. If batchable, all batchable
  subscriptions targeting the same branch/repo combination will share a PR.
  *Note: Batchable subscriptions are currently unsupported in darc*
- A set of auto merge policies, if the subscription is not batchable.  If batchable,
  merge policies are set on a repository level rather than a per-subscription
  level, as they end up shared between several subscriptions. *Note: repository
  merge policies are currently unsupported in darc*
  

> `add-subscription` has two modes of operation

Reviewer: >operation [](start = 36, length = 9)

operation -> operation:

**Parameters**

- `-channel` - **(Required if -q is passed)** Name of channel that is the source of the subscription. For a
  list of channels, see [get-channels](#get-channels)
- `--source-repo` - **(Required if -q is passed)** Source repository for the subscription.  Builds of this
  repository that appear on the specified `--channel` will have matching outputs
  applied to the inputs (specified in eng/Version.Details.xml) of `--target-repo` and `--target-branch`.
- `--target-repo` - **(Required if -q is passed)** Target repository for the subscription.  Builds of
  `--source-repo` that appear on the specified `--channel` will have matching
  outputs applied to the inputs (specified in eng/Version.Details.xml) of this
  repo's `--target-branch`
- `--target-branch` - **(Required if -q is passed)** Target branch for the subscription. Builds of
  `--source-repo` that appear on the specified `--channel` will have matching
  outputs applied to the inputs (specified in eng/Version.Details.xml) on this
  branch of `--target-repo`.
- `--update-frequency` - **(Required if -q is passed)** Frequency of updates. Valid values are: 'none',

>   'everyDay', or 'everyBuild'.  Every day is applied at 5am.  Subscriptions with

Reviewer: >Every day [](start = 32, length = 9)

Every day -> 'everyDay'

  'none' frequency can still be triggered using [trigger-subscriptions](#trigger-subscriptions)
- `--all-checks-passed` - Merge policy. A PR is automatically merged by Maestro++ if there is at least one
  check and all are passed. Optionally provide a comma separated list of
  ignored check with --ignore-checks.

> - `--ignore-checks` - Merge policy. A For use with --all-checks-passed. A set of checks that are

Reviewer: >A For [](start = 36, length = 5)

A For -> For(?)

  ignored. Typically, in github repos the "WIP" and "license/cla" checks are ignored.
- `--no-extra-commits`- Merge policy. A PR is automatically merged if no non-bot commits exist in the PR.

> - `--require-checks` - Merge policy. A PR is automatically merged if the specified checks are passed. Provide a comma separate list of required checks.

Reviewer: >`--require-checks [](start = 2, length = 17)

Split

- `-q, --quiet` - Non-interactive mode (requires all elements to be passed on the command line).

**Sample**:
```

> PS D:\enlistments\arcade-services> darc add-subscription --channel ".NET Tools - Latest" --source-repo https://github.com/dotnet/arcade --target-repo https://dev.azure.com/dnceng/internal/_git/dotnet-optimization --target-branch master --update-frequency everyDay --all-checks-passed -q

Reviewer: >PS D:\enlistments\ [](start = 0, length = 18)

Split

- [trigger-subscriptions](#trigger-subscriptions)
- [get-channels](#get-channels)

### **`authenticate`**

Set up your darc client so that the PAT or password inputs do not need to be
passed on each command invocation.  This command opens up an editor form with
various password settings. These values are overridden by the `--password`,
`--bar-uri`, `--azdev-pat` and `--github-pat` settings common to all commands.

See [Setting up Your Darc Client](#setting-up-your-darc-client) for more
information.

**Parameters**

None.

**Sample**
```
PS D:\enlistments\arcade> darc authenticate

(opens in editor)

# Create new BAR tokens at https://maestro-prod.westus2.cloudapp.azure.com/Account/Tokens
bar_password=***
# Create new GitHub personal access tokens at https://github.com/settings/tokens (no auth scopes needed)
github_token=***
# Create new Azure Dev Ops tokens at https://dev.azure.com/dnceng/_details/security/tokens (code read scope needed)
azure_devops_token=***
build_asset_registry_base_uri=https://maestro-prod.westus2.cloudapp.azure.com/

# Storing the required settings...
# Set elements above depending on what you need

```

### **`delete-channel`**

> Delete a channel. This channel must be in use by any subscriptions.

Reviewer: >This channel must be in use by any subscriptions. [](start = 18, length = 49)

So, if I create a channel that I realize after that I won't use so it is not used by any subscription, I won't be able to delete it?

Author: must not be.

*This is not a typical operation and you should consult with the (`@dnceng`)
engineering team before doing so.*

**Parameters**

- `-n, --name` - **(Required)** Name of channel to delete.

**Sample**:
```
PS D:\enlistments\arcade> darc delete-channel --name "Foo"

Successfully deleted channel 'Foo'.
```

**See also**:
- [add-channel](#add-channel)
- [get-channels](#get-channels)

### **`delete-default-channel`**

Deletes a default channel mapping. Deleting will not affect any existing builds,

> but new builds of the specified repos will will not be applied to the target

Reviewer: >will [](start = 43, length = 4)

Dup

channel.

You can obtain a list of current default channel mappings with
[get-default-channels](#get-default-channels)

- `--channel` - **(Required)** Name of channel that builds of 'repository' and 'branch' should not apply to.
- `--branch` - **(Required)** Repository that should have its default association removed.
- `--repo` - **(Required)** Branch that should have its default association
  removed.

**Sample**
```

> PS D:\enlistments\arcade> darc delete-default-channel --channel ".Net Core 3 Dev" --branch refs/heads/master --repo https://github.com/dotnet/arcade

Reviewer: >PS D:\enlis [](start = 0, length = 11)

Break into more lines

  is passed and the build contains non-shipping binaries). Under these
  folders will be two additional folders: 'assets' and 'packages'. Assets
  contains all non-package outputs, while 'packages' contains all nuget
  packages.
  
**Parameters**

- `-i, --id` - BAR ID of build to download. For information on locating the
  "root build", see [Gathering a build drop](#gathering-a-build-drop)
- `-r, --repo` - If set, gather a build drop for a build of this repo. Requires
  --commit. For information on locating the
  "root build", see [Gathering a build drop](#gathering-a-build-drop)
- `-c, --commit` - Branch, commit or tag to look up and gather a build drop for.
  For information on locating the
  "root build", see [Gathering a build drop](#gathering-a-build-drop)
- `-o, --output-dir` - **(Required)** Output directory to place build drop.
- `-f, --full` - Gather the full drop (build and all input builds).
- `-s, --separated` - Separate out each source repo in the drop into separate directories.
- `--continue-on-error` - Continue on error rather than halting.  Allows for
  gathering drops in cases where some outputs might not be able to be
  downloaded.

> - `--non-shipping`` - (Default: true) Include non-shipping assets.

Reviewer: >`` [](start = 17, length = 2)

Remove one '`'

- `--overwrite` - Overwrite existing files at the destination.
- `--dry-run` - Do not actually download files, but print what we would do.
- `--include-toolset` - Include toolset dependencies.

**Sample**:

Isolated drop:

```
PS C:\enlistments\core-sdk> dotnet C:\enlistments\arcade-services\artifacts\bin\Microsoft.DotNet.Darc\Debug\netcoreapp2.1\Microso

> ft.DotNet.Darc.dll gather-drop --output-dir C:\scratch\core-sdk-drop --commit 465a336c7a5ca3af2f6cf5172ddc0ebde620803b --repo https://github.com/dotnet/core-sdk

Reviewer: >ft.D [](start = 0, length = 4)

Break into lines (here and elsewhere applicable)

Looking up build 2038 in Build Asset Registry...
Finding build for Microsoft.DotNet.Cli.Runtime@3.0.100-preview.19075.1...
Looking up Microsoft.DotNet.Cli.Runtime@3.0.100-preview.19075.1 in Build Asset Registry...
Looking up build 1946 in Build Asset Registry...
Finding build for Microsoft.NET.Sdk@3.0.100-preview.19075.2...
Looking up Microsoft.NET.Sdk@3.0.100-preview.19075.2 in Build Asset Registry...
Looking up build 1931 in Build Asset Registry...
Finding build for Microsoft.Build@16.0.0-preview.386...
Looking up Microsoft.Build@16.0.0-preview.386 in Build Asset Registry...
Looking up build 2010 in Build Asset Registry...
Finding build for Microsoft.NETCore.Compilers@3.0.0-beta3-19067-14...
Looking up Microsoft.NETCore.Compilers@3.0.0-beta3-19067-14 in Build Asset Registry...
Looking up build 1598 in Build Asset Registry...
Finding build for Microsoft.NET.Sdk.Razor@3.0.0-preview-19079-02...
Looking up Microsoft.NET.Sdk.Razor@3.0.0-preview-19079-02 in Build Asset Registry...
Looking up build 2015 in Build Asset Registry...
Finding build for Microsoft.NET.Sdk.WindowsDesktop@3.0.0-preview-27329-5...

> Lo

Reviewer: >displayed.  [](start = 0, length = 11)

Remove and add new break.

> Lo

Reviewer: >s [](start = 62, length = 1)

Add '.'

> Lo

Reviewer: >/ [](start = 69, length = 1)

/ -> .

## Index
- [Scenarios](#scenarios)
  - [Setting up your darc client](#setting-up-your-darc-client)
  - [Adding dependencies to a repository](#adding-dependencies-to-a-repository)
  - [Updating dependencies in your local repository](#updating-dependencies-in-your-local-repository)
  - [Removing dependencies from a repository](#removing-dependencies-from-a-repository)
  - [Changing a dependency's type](#changing-a-dependencys-type)
  - ['Pinning' dependencies so they do not update](#pinning-dependencies-so-they-do-not-update)
  - [Adding dependency flow](#adding-dependency-flow)
  - [Halting and restarting dependency flow](#halting-and-restarting-dependency-flow)
  - [Viewing the dependency graph](#viewing-the-dependency-graph)
  - [Gathering a build drop](#gathering-a-build-drop)
  
- [Command Reference](#command-reference)
  - [Common Parameters](#common-parameters)
  - [add-channel](#add-channel) - Creates a new channel.
  - [add](#add) - Add a new dependency to Version.Details.xml.
  - [add-default-channel](#add-default-channel) - Add a channel that a build of a branch+repository is automatically applied to.
  - [add-subscription](#add-subscription) - Add a new subscription.

>   - [authenticate](#authenticate) - Stores the VSTS and GitHub tokens required for remote operations.

Reviewer 2: ```suggestion
  - [authenticate](#authenticate) - Stores the Azure DevOps and GitHub tokens required for remote operations.
```

  
  YAML format for interactive mode:
  ```
   - Name: AllChecksSuccessful
     Properties:
       ignoreChecks:
       - WIP
       - license/cla
       - <other check names>
  ```
   
- RequireChecks - Require that a specific set of checks pass. The check name
  corresponds to the string that shows up in GitHub/Azure DevOps.
  
  YAML format for interactive mode:
  ```
   - Name: RequireChecks
     Properties:
       checks:
       - MyCIValidation
       - CI
       - <other check names>
  ```
   
- NoExtraCommits - If additional non-bot commits appear in the PR, the PR should not be merged.

  YAML format for interactive mode:
  ```
   - Name: NoExtraCommits
  ```

**See also**:
- [delete-subscription](#delete-subscription)
- [get-subscriptions](#get-subscriptions)
- [trigger-subscriptions](#trigger-subscriptions)
- [get-channels](#get-channels)

### **`authenticate`**

Set up your darc client so that the PAT or password inputs do not need to be
passed on each command invocation.  This command opens up an editor form with
various password settings. These values are overridden by the `--password`,
`--bar-uri`, `--azdev-pat` and `--github-pat` settings common to all commands.

> 

Reviewer 2: I suggest highlighting the required scopes for the PATs. They are mentioned in the example below, but up here would be more noticeable.
